src/test/app/ReducedOffer_test.cpp

```diff
@@ -0,0 +1,626 @@
+    {
+        testcase("exercise partial cross new XRP/IOU offer Q change");
+
+        using namespace jtx;
+
+        auto const gw = Account{"gateway"};
+        auto const alice = Account{"alice"};
+        auto const bob = Account{"bob"};
+        auto const USD = gw["USD"];
+
+        // Make one test run without fixReducedOffersV1 and one with.
+        for (FeatureBitset features :
+             {supported_amendments() - fixReducedOffersV1,
+              supported_amendments() | fixReducedOffersV1})
+        {
+            Env env{*this, features};
+
+            // Make sure none of the offers we generate are under funded.
+            env.fund(XRP(10'000'000), gw, alice, bob);
+            env.close();
+
+            env(trust(alice, USD(10'000'000)));
+            env(trust(bob, USD(10'000'000)));
+            env.close();
+
+            env(pay(gw, bob, USD(10'000'000)));
+            env.close();
+
+            // Lambda that:
+            //  1. Exercises one offer pair,
+            //  2. Collects the results, and
+            //  3. Cleans up for the next offer pair.
+            // Returns 1 if the crossed offer has a bad rate for the book.
+            auto exerciseOfferPair =
+                [this, &env, &alice, &bob](
+                    Amounts const& inLedger,
+                    Amounts const& newOffer) -> unsigned int {
+                // Put inLedger offer in the ledger so newOffer can cross it.
+                std::uint32_t const aliceOfferSeq = env.seq(alice);
+                env(offer(alice, inLedger.in, inLedger.out));
+                env.close();
+
+                // Now alice's offer will partially cross bob's offer.
+                STAmount const initialRate = Quality(newOffer).rate();
+                std::uint32_t const bobOfferSeq = env.seq(bob);
+                STAmount const bobInitialBalance = env.balance(bob);
+                STAmount const bobsFee = drops(10);
+                env(offer(bob, newOffer.in, newOffer.out, tfSell),
+                    fee(bobsFee));
+                env.close();
+                STAmount const bobFinalBalance = env.balance(bob);
+
+                // alice's offer should be fully crossed and so gone from
+                // the ledger.
+                {
+                    Json::Value aliceOffer =
+                        ledgerEntryOffer(env, alice, aliceOfferSeq);
+                    if (!BEAST_EXPECT(
+                            aliceOffer.isMember(jss::error) &&
+                            aliceOffer[jss::error].asString() ==
+                                "entryNotFound"))
+                        // If the in-ledger offer was not consumed then further
+                        // results are meaningless.
+                        return 1;
+                }
+                // bob's offer should be in the ledger, but reduced in size.
+                unsigned int badRate = 1;
+                {
+                    Json::Value bobOffer =
```

Comment:
Consider making lines 114-144 into a function. There is another similar block of code.

Reply:
Were you planning to add a function for this? It's not in the latest commit. I'm fine if you decided not to. Just double checking since you marked the comment with :+1:.